File: src/leetcode/python/easy/RemoveDuplicatesFromSortedList_83.py
class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None


class RemoveDuplicatesFromSortedList_83(object):
    def deleteDuplicates(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        cur = head
        while cur:
            while cur.next and cur.next.val == cur.val:
                cur.next = cur.next.next
            cur = cur.next
        return head

# old version:
#
# if head is None:
#     return None
# if head.next is None:
#     return head
# pre = nxt = head
# s = set()
# while nxt:
#     if nxt.val not in s:
#         s.add(nxt.val)
#         pre = nxt
#     else:
#         pre.next = nxt.next
#     nxt = nxt.next
# return head

File: src/leetcode/java/RemoveDuplicatesFromSortedList_83.java
package leetcode.java;

import java.util.HashSet;
import java.util.Set;

public class RemoveDuplicatesFromSortedList_83 {
  public ListNode deleteDuplicates(ListNode head) {
    if (head == null) return null;
    if (head.next == null) return head;
    Set<Integer> dup = new HashSet<>();
    ListNode pre = head, nxt = head;
    while (nxt != null) {
      if (dup.add(nxt.val)) {
        pre = nxt;
      } else {
        pre.next = nxt.next;
      }
      nxt = nxt.next;
    }
    return head;
  }

  class ListNode {
    int val;
    ListNode next;

    ListNode(int x) {
      val = x;
    }
  }
}
